Network binding: Add arping support, iOS port knock, async cache instead of GUI block. Add test package.

davidgraeff opened the pull request:
Following features are added to the network binding:
* Move not exported packages to the "internal" namespace, except the "constants" class.
* Not blocking the main thread when the GUI asks for new values.
* The external native tool "arping" can be used for presence detection.
* Use port knocking for iOS devices for the arp ping method.
* Add a new thing type to distinguish between a "Service device", identified by a TCP connection attempt and DHCP traffic observation and a "Pingable device". The latter is identified by ICMP pings, arp pings and DHCP traffic observation.
* Some thing configurations are moved to be binding configurations now:
  - Use DHCP sniffing (default is on) 
  - ARP ping tool path (default is "arping", but can be any absolute path to the tool)
  - Cache device state (default is 2000ms)
  - Allow system pings (default is on)
* DHCP sniffing will use port 6767 if port 67 is not bindable. This way the user can at least
  configure a port forwarding to use this function. In the previous implementation the user
  could not use this functionality if elevated privileges didn't work for him.
* Add unit tests for every class.

## Breaking change
The old thing "device" is now "pingdevice". This is not a real breaking change though, as all old configurations should still work because I map "device" to "pingdevice" in the handler factory.

## Auto discovery
![screenshot from 2017-06-18 17-57-54](https://user-images.githubusercontent.com/66436/27259168-d418368c-544f-11e7-8aa8-36b81415419d.png)

The discovery searches for pingable devices as before but additionally also for service devices. At the moment it is only scanning for devices with open port 554 (windows file sharing).

## Less configuration
![screenshot from 2017-06-18 17-58-24](https://user-images.githubusercontent.com/66436/27259172-eb36b140-544f-11e7-8ba8-9390b68cb001.png)

Native ping, arp ping and dhcp sniffing are enabled by default. A feature detection is performed on thing creation. The user can disable native ping (and use the java .isReachable() method instead) in the binding configuration. ARP pings will be performed if the "arping" tool can be executed. DHCP sniffing will be enabled if the DHCP port can be opened.

## More information without looking into the log
![screenshot from 2017-06-18 17-53-50](https://user-images.githubusercontent.com/66436/27259129-25281142-544f-11e7-9133-d3d039f20fb9.png)
The things provide more information to the user.
* The DHCP sniffing status is communicated through a thing property, including the exception error message (which indicates if the port is unavailable because of missing access rights).
* The last successful presence detection type(DHCP, ARP ping, ICMP ping, TCP) is available as a property.
* The ARP ping tool status is available and any miss-configuration of the external tool path is spotted immediately.

## Cache
The result of the device presence detection is cached for a small amount of time (2000ms, configurable). I considered the use of the new ExpiringCache class, but I'm not entirely happy with the API because I'm in need of a more asynchronous approach here. The goal is to not block the framework thread in **handleCommand(..)**. At the moment this is realized with a new ExpiringCacheAsync class.

Fixes #2365
Fixes #1579
Fixes #1156
Fixes #2271 
Fixes #2389 
Fixes #2037

martinvw commented:
Wow the PR-message itself looks perfect, will try to review ASAP :-)

cweitkamp commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/cache/ExpiringCacheAsync.java:
```diff
@@ -0,0 +1,116 @@
+    public void refreshValue(ExpiringCacheValue<V> callback) {
+        expiresAt = 0;
+        waitingCacheCallbacks.add(callback);
+        // TODO prevent calling this more often than expire time
```
This should be done. ;-)

cweitkamp commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/cache/ExpiringCacheAsync.java:
```diff
@@ -0,0 +1,116 @@
+     * @param expiry the duration in milliseconds for how long the value stays valid
+     * @param cacheUpdater The cache will use this callback if a new value is needed.
+     */
+    public ExpiringCacheAsync(long expiry, ExpiringCacheUpdate cacheUpdater) {
```
Null check for `cacheUpdater` to prevent NPE in [refreshValue](https://github.com/openhab/openhab2-addons/pull/2380/files#diff-e4d04165f1465214d41ca00e2de0a61cR105)?

cweitkamp commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/cache/ExpiringCacheAsync.java:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return the value
+     */
+    public void getValueAsync(ExpiringCacheValue<V> callback) {
```
Shall we call it `getValue` to be complient with the `ExpiringCache` class?

IMHO the `callback` should be handed over in the constructor and stored in a private property. This makes it possible to handle the callback completely internal without using a public `setValue `method. wdyt?

davidgraeff replied:
Not a good idea, at least in my use case I need multiple consumers, not just one

cweitkamp replied:
Hm, ... then you should use multiple instances of the `ExpiryCache` or an `ExpiringCacheMap` to handle multiple values. Each cached value should only be updated in one specific way. If you assign different `Consumer` for one value there will be a risk to retrieve a wrong value from the cache.

davidgraeff replied:
ExpiryCache(Async) contains one value. When getValue() is called and the value is not expired, all consumers of that API retrieve the same value. If the value is expired they all go in a callback list. As soon as setValue() is called, all consumers get notified with the same value. Do I miss something?

cweitkamp replied:
Understood. You are right. I misinterpreted the situation.

cweitkamp commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/service/NetworkService.java:
```diff
@@ -29,159 +31,276 @@
-                logger.debug("ping program was interrupted");
+    public void startAutomaticRefresh(ScheduledExecutorService scheduledExecutorService) {
+        cancelRefreshJob();
+        refreshJob = scheduledExecutorService.scheduleWithFixedDelay(new Runnable() {
```
If I understand the code correctly, the `ExpiringCacheAsync` calls this method every time the value should be refreshed. So we can use a one-shot [schedule](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ScheduledExecutorService.html#schedule(java.lang.Runnable,%20long,%20java.util.concurrent.TimeUnit)) here.

davidgraeff replied:
If I understand you correctly, we use a one shot in "startAutomaticRefresh()" and then a one shot in each "requestCacheUpdate()" call? I'm not sure if we gain a lot by this change.

cweitkamp replied:
Ok, maybe I should ask my question from another point of view: Why should the cache refresh call the `networkService.startAutomaticRefresh()` method to stop and restart all scheduled jobs? IMHO it only needs to execute one special job to refresh the requested value. Doesn't it?

davidgraeff replied:
Kai asked basically the same question (https://github.com/openhab/openhab2-addons/pull/1989#discussion_r122158419) ("Do you really think that messing with the scheduled job is the right approach?")

My response was: "Setting a new timeout for the scheduled job makes sense to me, yes. To my understanding the user expresses the wish that as soon as he knows about the (non-)reachable state of a device, he wants the next check to happen after the given period of time. So if the GUI causes a state refresh earlier in time, the next periodic check can be postponed."

davidgraeff commented:
@cweitkamp Thanks for the review, I will adapt the code based on your comments. I rather not want this to be a PR blocker though. I'm not complaining but my average merge time frame here is 5 month, I don't want this class to delay the PR by waiting for the ESH core team. I will happily adapt the code in a later PR, as soon as ESH provides a comparable solution, of course.

I'm not sure if we can discuss it here, or should head over to the community.openhab or ESH forum (I don't like the ESH forum that much though, who uses BBcode nowadays).

I did some research and I recon that the existing ExpiringCache together with the Guava ListenableFuture may provide a similar interface to my ExpiringCacheAsync class. Maybe we can also come up with a solution to just extend ExpiringCache?

cweitkamp commented:
@davidgraeff I don't want to block your PR. I just thought about a general improvement.

I will open an issue in the ESH Repo to start a discussion on it.

I think we should try to solve this without using Guava. We implemented the `ExpiringCache` to get rid of this dependency. In ESH we are currently using Guava 10.1 and there are conflicts in updating to newer versions (see [#2801](https://github.com/eclipse/smarthome/issues/2801) why it is not a good idea).

davidgraeff commented:
Just a general question regarding this binding: Should the "thing" go offline, it the target device is not reachable? At the moment it stays online except if there is a configuration error and only the channel "online" goes ON/OFF.

martinvw commented:
> Just a general question regarding this binding: Should the "thing" go offline, it the target device is not reachable? At the moment it stays online except if there is a configuration error and only the channel "online" goes ON/OFF.

Good question, @kaikreuzer, wdyt?

kaikreuzer commented:
The way as we treat it now (i.e. by having a channel "online"), it actually means that the Thing is not the remote device, but rather the "pinging service for that device" - and this one should stay ONLINE as long as it is working correctly.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/service/NetworkService.java:
```diff
@@ -29,159 +31,276 @@
+     * Return true if this device is recognised as IOS device
+     */
+    public boolean isIOSdevice() {
+        return true;
```
Are you sure this should always be true

davidgraeff replied:
I have no solution for this yet, maybe this should a configuration option for the binding or the thing. I'm pretty sure iOS devices can be recognized somehow, I just don't have one.

martinvw replied:
Should we then leave out this method for now or add a todo. Note maybe you can base it on the MAC address but that is quite an extended list.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/service/NetworkUtils.java:
```diff
@@ -56,13 +64,37 @@
+     *
+     * @return Set of interface names
+     */
+    public static TreeSet<String> getInterfaceNames() {
```
Can it return a `Set ` or maybe `SortedSet `or should it really be a `TreeSet`, see also next line

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/service/NetworkUtils.java:
```diff
@@ -146,25 +178,125 @@ public static final long inetAddress2Long(InetAddress addr) {
+            return true;
+        } catch (SocketTimeoutException ignored) {
+            return false;
+        } catch (ConnectException e) {
```
Since you ignore them all (should you really not `trace `or `debug `log them?) you can also call them all  `ignored`

davidgraeff replied:
We actually expect one of those exceptions if the device is not reachable. I would not log this event here.

martinvw replied:
Ok

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/service/NetworkUtils.java:
```diff
@@ -146,25 +178,125 @@ public static final long inetAddress2Long(InetAddress addr) {
+     * @param logger A slf4j logger instance to log IOException
+     * @return
+     */
+    public static boolean servicePing(String host, int port, int timeout, Logger logger) {
```
Why pass the logger you can and imho should use a static logger in this case. Otherwise you lose the context where this logging came from.

davidgraeff replied:
I thought static loggers are a no-go? I can use a static one here of course.

martinvw replied:
> As we are in a dynamic OSGi environment, loggers should be non-static, when ever possible and have the name logger.

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/service/NetworkService.java:
```diff
@@ -29,159 +31,276 @@
+                                    timeoutInMS)) {
+                                if (!reachable) {
+                                    reachable = true;
+                                    updateListener.updatedDeviceState(
```
Agreed

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/NetworkHandlerFactory.java:
```diff
@@ -10,34 +10,92 @@
+
+        try {
+            if (thingTypeUID.equals(PING_DEVICE)) {
+                return new NetworkHandler(thing, false, allowSystemPings, allowDHCPlisten, cacheDeviceStateTimeInMS,
```
I can use the builder pattern here

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/discovery/NetworkDiscoveryService.java:
```diff
@@ -23,25 +28,47 @@
+    static final int PING_TIMEOUT_IN_MS = 500;
+    // TCP port 548 (Apple Filing Protocol (AFP))
+    // TCP port 554 (Windows share / Linux samba)
+    // TCP port 1025 (XBox / MS-RPC)
```
I thought about adding more port scanning to detect devices, yes. I'm just not sure how I should handle the situation when a device is "detected" multiple times with different ports

martinvw replied:
If you generate the same thingUID it should not be duplicated but the last one should 'win'

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/discovery/NetworkDiscoveryService.java:
```diff
@@ -51,15 +78,71 @@ protected void startScan() {
+                for (final String interfaceName : interfaceNames) {
+                    executorService.execute(() -> {
+                        try {
+                            NetworkUtils.wakeUpIOS(destination);
```
It doesn't harm to send a udp packet before sending an arp packet. But you are right, the method `isIOSdevice()` should be used before calling wakeUpIOS().

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/discovery/NetworkDiscoveryService.java:
```diff
@@ -23,25 +28,47 @@
  * the current Network. It uses every Network Interface which is connected to a network.
+ * It tries common TCP ports to connect to, ICMP pings and ARP pings.
  *
+ * @author David Graeff
```
Most of time I have seen it other way round. The last person that changes a file comes first (makes actually sense if you want to contact a person for asking for implementation details).

martinvw replied:
Ok, then just add on the top you reasoning sounds valid

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/NetworkBindingConstants.java:
```diff
@@ -32,6 +32,7 @@
     // List of all Channel ids
     public static final String CHANNEL_ONLINE = "online";
     public static final String CHANNEL_TIME = "time";
```
yes, that needs to be renamed, I agree.

davidgraeff commented on addons/binding/org.openhab.binding.network/OSGI-INF/NetworkHandlerFactory.xml:
```diff
@@ -9,7 +9,10 @@
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0"
+    immediate="true"
+    configuration-policy="optional"
+    name="org.openhab.binding.network">
```
I've seen this in other binding updates, that the component name of the handlerFactory is shortened to the bindings package name.

martinvw replied:
But most important does it work, can you change the config as expected?

davidgraeff commented on addons/binding/org.openhab.binding.network/ESH-INF/thing/thing-types.xml:
```diff
@@ -3,71 +3,118 @@
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
         xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd"><!--Network Binding-->
-    <thing-type id="device">
```
That's true. Most people use the "ping" device, I propose therefore to add a "device" think mapping to "pingdevice" in the handler factory. It will break for people who are using the "port" parameter, but it was a bad idea in the first place to combine this in one thing. There was an issue I linked to in the PR description: When you use the port parameter, all ping parameters are not used at all. That confused people in the forum and we really should fix this now than later.

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/handler/NetworkHandler.java:
```diff
@@ -75,46 +99,68 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+            networkService.setHostname(String.valueOf(conf.get(PARAMETER_HOSTNAME)));
+        } catch (UnknownHostException e) {
+            logger.error("Configuration for hostname is faulty", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, e.getLocalizedMessage());
```
This is a good question. I would say it is a configuration error. It is actually an additional nice-to-have feature to allow people in the first place to use a dns name instead of an IP. The dns name should be at least resolvable in the startup process of the binding. Especially because the dns resolving does not talk to the target device but to a dns server.

martinvw replied:
I have a dynamic dns resolver in my home so that is why I asked 😆 

So if a device is off and the lease expired my gateway might not know about it

davidgraeff replied:
Most users won't have this configuration, can we think about this in another PR please? (Can you open an issue after merge?)

martinvw replied:
ok, but it was also just guess so I'm not even sure that it fails the way I would expect. I will try to test it first.

davidgraeff commented:
I addressed all issues including the wish to not change the scheduled job, but issue a new discovery instead. I have also moved non exported packages to the "internal" namespace (like in #2393).

Some files have changed considerably and need a complete new review I'm afraid. The affected classes are:
* PresenceDetectionService
* NetworkDiscoveryService

Because the cache moved from the handler to the service, the following class is affected too:
* NetworkHandler

Some new classes/interfaces are:
* DHCPListener
* NetworkHandlerBuilder
* PresenceDetectionValue

davidgraeff commented:
I will add a network.test the next days, this PR is therefore back to a WIP state.

davidgraeff commented:
@martinvw I have added unit tests for all classes. I don't now though why jenkins is failing on this PR.

martinvw commented on addons/binding/org.openhab.binding.network.test/pom.xml:
```diff
@@ -0,0 +1,126 @@
+  <parent>
+    <groupId>org.openhab.binding</groupId>
+    <artifactId>pom</artifactId>
+    <version>0.9.0-SNAPSHOT</version>
```
This is the problem

martinvw commented:
Force rebuild

martinvw commented:
Can you take a look at the travis problem?

davidgraeff commented:
Apparently the tests fail or don't build.
The eclipse setup for eclipse oxygen is highly fragile right now, I have to decide if I want to develop for ESH or OH2 at the moment (jdt.core/ui conflicting version problems). I can't execute OH bindings as I'm focused on the ESH Mqtt implementation these days.

davidgraeff commented:
@martinvw : The manifest listed some old ESH bundles. Fixed. I squashed everything already. The PR should be ready then :)

ThomDietrich commented:
@davidgraeff if a user wants to use the "last seen" information in a rule, a channel bound to an DateTime Item would be the way to go. Could you elaborate on ["I see no point in making it available as a channel."](https://github.com/openhab/openhab2-addons/issues/2039)

davidgraeff commented:
@ThomDietrich You can use thing properties as well in rules, can't u? If not, it's time to add this to the core I guess :D

ThomDietrich commented:
I'm not aware of that. Of course that would be amazing.
The only thing I learned about only today is https://github.com/eclipse/smarthome/pull/3001

How do you differentiate property vs channel? The idea of a last seen information can be represented as the "last updated" property of a Thing but could also be the "last seen" data channel of a Thing. Is there any clear distinction?

@kaikreuzer

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/dhcp/DHCPPacket.java:
```diff
@@ -88,6 +88,18 @@
     private Map<Byte, byte[]> options;
 
     /**
+     * Package private constructor for test suite.
```
Good that you add the comment but it would be great if it can be avoided for example by having a testbuilder in the test package.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/dhcp/DHCPPacketListenerServer.java:
```diff
@@ -0,0 +1,112 @@
+    DHCPPacketListenerServer(IPRequestReceivedCallback listener) throws SocketException, BindException {
+        this.listener = listener;
+        try {
+            dsocket = new DatagramSocket(null);
```
It would be clearer to seperate this body an the body's of the catch with a parameter for the port number.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/dhcp/DHCPPacketListenerServer.java:
```diff
@@ -0,0 +1,112 @@
+
+        Byte dhcpMessageType = request.getDHCPMessageType();
+
+        if (dhcpMessageType == null || dhcpMessageType != DHCPPacket.DHCPREQUEST) {
```
Is the second part already enough?

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/dhcp/DHCPPacketListenerServer.java:
```diff
@@ -0,0 +1,112 @@
+                }
+
+            }
+        } catch (IOException e) {
```
Will this catch any exception I assume they will be caught by the inner catch or do they have a different source.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/dhcp/DHCPPacketListenerServer.java:
```diff
@@ -0,0 +1,112 @@
+                try {
+                    receivePacket(new DHCPPacket(packet), packet.getAddress());
+                } catch (BadPacketException | IOException e) {
+                    return;
```
Since you return here I think you are better of with single try-catch instead of the current nested one.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java:
```diff
@@ -0,0 +1,304 @@
+     */
+    public LinkedHashSet<String> getNetworkIPs(Set<String> interfaceIPs) {
+        LinkedHashSet<String> networkIPs = new LinkedHashSet<String>();
+
```
Is there maybe some interface here which can be returned instead of the implementation class.

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java:
```diff
@@ -0,0 +1,304 @@
+    public LinkedHashSet<String> getNetworkIPs(Set<String> interfaceIPs) {
+        LinkedHashSet<String> networkIPs = new LinkedHashSet<String>();
+
+        for (Iterator<String> it = interfaceIPs.iterator(); it.hasNext();) {
```
Why not iterate this normally?

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/utils/NetworkUtils.java:
```diff
@@ -0,0 +1,304 @@
+     */
+    public boolean servicePing(String host, int port, int timeout) throws IOException {
+        try {
+            SocketAddress socketAddress = new InetSocketAddress(host, port);
```
I assume that we close in most cases in which we open it, but a try with resource could work out well here.

davidgraeff replied:
Unfortunately the InetSocketAddress/SocketAddress do not implement the Closable interface and cannot be used for try with resource.

martinvw replied:
Okay wrong line 😉

martinvw commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/internal/dhcp/DHCPListenService.java:
```diff
@@ -0,0 +1,69 @@
+                IPRequestReceivedCallback listener = registeredListeners.get(ipAddress);
+                if (listener != null) {
+                    instance.logger.info("DHCP request for registered address: {}", ipAddress);
+                    listener.dhcpRequestReceived(ipAddress);
```
This comment is intended to be about the logger, I might have misplaced it.

davidgraeff replied:
I need a singleton for DHCP (only one additional socket should listen on the udp port). But I didn't want to make the `DHCPPacketListenerServer` class a singleton because it is harder to test singletons. Therefore this new class `DHCPListenService` creates a single object of the server and allows to register/unregister callbacks. If you have another design idea, please let me now.

martinvw replied:
Sorry nothing wrong with this. The comment above was not only misplaces but als duplicate, I had some WiFi issues on the camping 😉

ThomDietrich commented on addons/binding/org.openhab.binding.network/ESH-INF/thing/thing-types.xml:
```diff
@@ -1,73 +1,119 @@
+<thing:thing-descriptions bindingId="network" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="http://eclipse.org/smarthome/schemas/thing-description/v1.0.0 http://eclipse.org/smarthome/schemas/thing-description-1.0.0.xsd"><!--Network Binding -->
+	<thing-type id="pingdevice">
```
To aid in the release notes process, all breaking changes should be highlighted in the first posting of the PR?

davidgraeff replied:
It is only a breaking change if someone used the port parameter, because I map the thingID "device" to "pingdevice" in the handler factory. And because we had two issues about the non working port parameter it is actually not a breaking change at all.

martinvw replied:
Great solution, I suppose this does definitely work for the textual configuration, did you also validate it for the paper ui configuration, I fear that you might not be able to change or remove the existing 'device' item.

martinvw replied:
Where you able to validate this?

davidgraeff replied:
It works. I stopped OH, renamed pingdevice in the jsondb file to device, same for textual files and started OH again. The things are picked up as pingdevices.

martinvw replied:
Yes I already expected that, but can you modify and delete them via the paper ui.

davidgraeff replied:
Yes, I was able to remove the paperUI device.

martinvw replied:
So also modify them works? I would suppose some crucial information for that should be missing, or the paper ui has some build in magic support for this work around.

If that is the case I would say this all perfect, except for the failing build. If you and @kaikreuzer could decide about that, maybe @htreu can help you / us out?

claell commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/discovery/NetworkDiscoveryService.java:
```diff
@@ -72,29 +145,72 @@ protected synchronized void stopScan() {
+                label = "Windows compatible device";
+                break;
+            case 1025:
+                label = "XBox compatibel device";
```
Also a small typo: "compatibel"

davidgraeff commented on addons/binding/org.openhab.binding.network.test/META-INF/MANIFEST.MF:
```diff
@@ -0,0 +1,24 @@
+ org.osgi.framework,
+ org.osgi.service.device,
+ org.slf4j
+Require-Bundle: org.junit,org.mockito,org.hamcrest
```
I'm puzzled to be honest, because I have done this in all my test bundles, and some got recently accepted to the ESH repository. And it is actually not made by me this way but by the skeleton script. Should I still change it? @martinvw

martinvw replied:
I do not see it in the source file of skeleton:

https://github.com/eclipse/smarthome/blob/master/tools/archetype/binding.test/src/main/resources/archetype-resources/META-INF/MANIFEST.MF

I think you should fix the other places where you used it as well.

davidgraeff replied:
@martinvw : The linked file contains "Require-Bundle: org.junit,org.mockito,org.hamcrest" :D

martinvw replied:
I might require some glasses, thanks for pointing it out!

I asked the writer of the specific changes and some other team members for some clarification.

martinvw replied:
It seems like the guidelines might require some updates, at least for now this guideline does not seem to apply to test packages 😔

htreu replied:
This is the discussion about the required bundles in test fragments: https://github.com/eclipse/smarthome/pull/3363#pullrequestreview-53592553

Guideline update is on the way.

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/cache/ExpiringCacheAsync.java:
```diff
@@ -0,0 +1,135 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.cache;
```
I was thinking about moving it to a separate bundle as it turns out to be a quite handy class for my bindings. It is actually always the case, that we don't want to block in `handleCommand` and return a cached value but at the same time retrieve a more recent one if the cached one is too old.

martinvw replied:
Just add it to esh somewhere close to the other caching solution.

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/NetworkHandlerFactory.java:
```diff
@@ -0,0 +1,98 @@
+ * @author Marc Mettke
+ */
+public class NetworkHandlerFactory extends BaseThingHandlerFactory {
+    private final Logger logger = LoggerFactory.getLogger(NetworkHandler.class);
```
Is it actually legit to just use `LoggerFactory.getLogger(getClass())` everywhere?

martinvw replied:
No it has some undesired behaviour with inheritance, the slf4j FAQ suggest a copy safe way but that is not yet used anywhere and pretty obscure. So just correct it 🙂

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/NetworkBindingConstants.java:
```diff
@@ -8,11 +8,12 @@
 
 import org.eclipse.smarthome.core.thing.ThingTypeUID;
 
+import com.google.common.collect.Sets;
```
Sets.newHashSet is used in every binding I have seen so far. To be honest I would not know how to accomplish something like `Set<ThingTypeUID> SUPPORTED_THING_TYPES_UIDS = Sets.newHashSet(DEV1, DEV2);` without this helper class in one line without writing an own helper that would need to be replicated in every binding. Is there any solution yet?

martinvw replied:
Maybe we should just wait for Java 9 they will finally solve this. I remember I saw someone else giving the comment, so maybe I remembered incorrect, maybe @sjka or @kaikreuzer know what I might have been reading.

sjka replied:
The following would be another way:

```java
new HashSet(Arrays.asList("1", "2", "3"))
```

It wasn't me giving the comment you are referring to. But I agree that we should avoid dragging in guava just for this. Indeed, many bindings are using it currently, the direction is towards avoiding it though. On devices with restricted resources, it's a pretty high price to pay just for saving a few characters when creating collections...

davidgraeff commented:
I addressed the mentioned problems except:
* If import com.google.common.collect.Sets should be used or not
* The test bundle doesn't work on travis and I have no idea why. It works locally.
* A testbuilder class for DHCPPacket.java. Personally I thing an empty package local constructor can't cause any damage.

martinvw commented:
> The test bundle doesn't work on travis and I have no idea why. It works locally.

We cannot merge without this solved, I won't  have time to look at this for at least one more week.

Also note that some of my replies might create some work as well, did you or can you test the compatibility of old paper ui configuration with this new binding.

davidgraeff commented:
Yeah, I test it tomorrow and make a pull request for the cache class to ESH. Can someone else with maven/pom knowledge look at the problem?

martinvw commented:
Sorry, I did not mention it yet but the error seems to be caused by the amazon dash button, I just found some possible cause.

triller-telekom commented:
Sorry my fault, I forgot one parameter in maven, will be fixed with: https://github.com/openhab/openhab-core/pull/174

davidgraeff commented:
Can someone trigger a rebuild later? :)

martinvw commented:
I am on it.

davidgraeff commented:
Still the same error, but that was expected: https://travis-ci.org/openhab/openhab2-addons/builds/260108039#L2471

martinvw commented on addons/binding/org.openhab.binding.network.test/pom.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
More important as far as I can see this does not follow the archetype, but I still do not have glasses so I might be wrong 😉

https://github.com/eclipse/smarthome/blob/master/tools/archetype/binding.test/src/main/resources/archetype-resources/pom.xml

davidgraeff replied:
The skeleton provided pom didn't work, so I tried different poms I have seen in other test bindings. The `<?xml` declaration should not be a problem though, it is an xml after all.

martinvw replied:
I mean the file not the xml declaration, I think you miss a lot of dependencies / plugins.

davidgraeff replied:
`mvn test` works. What is the command, that travis executes?

martinvw replied:
`mvn clean install -B -V`

Unfortunately just test does not execute the tests, this is because of the osgi maven test plugin we use 🙁

davidgraeff replied:
Someone should add this to the doc :D because the ESH doc clearly says "mvn test" ^^

martinvw replied:
It could be an openHAB issue, I noticed recently that `mvn test` did not do any tests when the build broke afer merging something which looked all green on travis.

It would be great if you could test it quickly by breaking one of your esh tests and running the test command. If it does not work for esh as well it would be great if you could great issue there otherwise create an issue in openHAB core and mention me.

davidgraeff replied:
You are right, I made one test fail on purpose, executed `mvn test` and it still says success. I will now test with an esh test.

davidgraeff commented:
So, we need to update the documentation then? My ESH `mvn test` is still running, it feels like it downloads the internet... twice.

martinvw commented:
And you have a conflict.

davidgraeff commented on addons/binding/org.openhab.binding.network/src/main/java/org/openhab/binding/network/toberemoved/cache/ExpiringCacheAsync.java:
```diff
@@ -0,0 +1,135 @@
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.network.toberemoved.cache;
```
Yes, I made a PR for ESH and this class will be removed in a later PR.

davidgraeff commented:
@kaikreuzer: If I would port the UDP and TCP binding functionality of OH1 over to OH2, would I add this to the network binding?

martinvw commented:
@kaikreuzer note that the conflict is resolved :-)

martinvw commented:
> I can remove the tests fragment though, so that this can be merged.

Did you remove the tests here? Because they are still not running properly on travis / maven

martinvw commented:
> Would be nice if someone else could look at this, I will not invest time in maven anymore. 

A I forgot about that request, my holidays are over so I might be able to find some time.

davidgraeff commented:
@martinvw : The tests are running fine locally. It is maven/tycho/pom/manifest magic that is failing.

martinvw commented:
> @martinvw : The tests are running fine locally. It is maven/tycho/pom/manifest magic that is failing.
I decided not to split this up, so we should fix this somehow for the merge to happen.

I will pick it up, I just forgot about it 😉

kaikreuzer commented:
> @kaikreuzer: If I would port the UDP and TCP binding functionality of OH1 over to OH2, would I add this to the network binding?

I am not sure how an UDP/TCP binding would look like in ESH. It is something pretty low-level, which cannot really benefit from any of the new ESH concepts - it's probably a similar discussion as for MQTT; the most important part would be a transport.tcp bundle, which brings useful functionality for connection handling that other bindings can then use. A "generic tcp binding" could then be offered on top of that. But feel free to create a separate issue (probably best directly at ESH as well) to discuss details.

martinvw commented:
And there is our solution :-)

https://github.com/meysholdt/Xpect/issues/165#issuecomment-76906598

davidgraeff commented:
@martinvw OMG, that is absolutely not obvious, thank you for finding the culprit. This should be documented, including the resulting error message.

martinvw commented:
@davidgraeff will you change it in your code? Please note that it was/is also complaining about the about.html not exposed from the test bundle.

martinvw commented:
@davidgraeff should we try to add it here in your opinion? http://www.eclipse.org/smarthome/documentation/development/testing.html

davidgraeff commented:
I would add it to https://www.eclipse.org/smarthome/documentation/development/testing.html#unit-tests.

Instead of:

> Each class inside the test folder, which has a public method with a @Test annotation will automatically be executed as a test.

It would read like:

> Each class **which class-name ends with ...Test** inside the test folder, which has a public method with a @Test annotation will automatically be executed as a test.

And would duplicate the information for  https://www.eclipse.org/smarthome/documentation/development/testing.html#osgi-tests:

> Eclipse SmartHome comes with an abstract base class JavaOSGiTest for OSGi tests. The base class sets up a bundle context and has convenience methods for registering mocks as OSGi services and the retrieval of registered OSGi services. **Public method with a @Test annotation will automatically be executed as an OSGi test, as long as the class-name ends with ...Test**. The following JUnit/Mockito test class shows how to test the ItemRegistry by providing a mocked ItemProvider.

davidgraeff commented:
And maybe we should add a section "common errors", where we list the runtime error and the solution for common errors

davidgraeff commented:
Ups, I haven't seen that you pushed, too. I amended and pushed over it :/

martinvw commented:
> Ups, I haven't seen that you pushed, too. I amended and pushed over it :/

No problem finding it out was harder than actually changing and pushing it 😄

martinvw commented:
@kaikreuzer all green now, please take a final look :-)

kaikreuzer commented:
Wow @davidgraeff, what an amazing work and what thorough reviews and discussions on that PR - I just quickly scanned over it and I fully trust @martinvw's approval :-)

claell commented:
Maybe the wrong place, but how long will it probably take until the new version is available on my OpenHAB installation?

ThomDietrich commented:
@claell https://community.openhab.org/t/frequently-asked-questions-faqs/17727#addons-1

claell commented:
Thanks, but what if I want to wait until it becomes available officially, so I can simply update it? Do I need to wait until OH 2.2 gets released?

davidgraeff commented:
@claell Yes

claell commented:
Alright
